x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/GetGlobalCheckpointsAction.java

```diff
@@ -0,0 +1,225 @@
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+            builder.startObject();
+            builder.array("global_checkpoints", globalCheckpoints);
+            return builder.endObject();
+        }
+    }
+
+    public static class Request extends ActionRequest implements IndicesRequest {
+
+        private final String index;
+        private final boolean waitForAdvance;
+        private final long[] currentCheckpoints;
+        private final TimeValue pollTimeout;
+
+        public Request(String index, boolean waitForAdvance, long[] currentCheckpoints, TimeValue pollTimeout) {
+            this.index = index;
+            this.waitForAdvance = waitForAdvance;
+            this.currentCheckpoints = currentCheckpoints;
+            this.pollTimeout = pollTimeout;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        public TimeValue pollTimeout() {
+            return pollTimeout;
+        }
+
+        public boolean waitForAdvance() {
+            return waitForAdvance;
+        }
+
+        public long[] currentCheckpoints() {
+            return currentCheckpoints;
+        }
+
+        @Override
+        public String[] indices() {
+            return new String[] { index };
+        }
+
+        @Override
+        public IndicesOptions indicesOptions() {
+            return IndicesOptions.strictSingleIndexNoExpandForbidClosed();
+        }
+    }
+
+    public static class TransportGetGlobalCheckpointsAction extends TransportAction<Request, Response> {
+
+        private final ClusterService clusterService;
+        private final NodeClient client;
+        private final IndexNameExpressionResolver resolver;
+
+        @Inject
+        public TransportGetGlobalCheckpointsAction(
+            final ActionFilters actionFilters,
+            final TransportService transportService,
+            final ClusterService clusterService,
+            final NodeClient client,
+            final IndexNameExpressionResolver resolver
+        ) {
+            super(NAME, actionFilters, transportService.getTaskManager());
+            this.clusterService = clusterService;
+            this.client = client;
+            this.resolver = resolver;
+        }
+
+        @Override
+        protected void doExecute(Task task, Request request, ActionListener<Response> listener) {
+            final ClusterState state = clusterService.state();
+            final Index index = resolver.concreteSingleIndex(state, request);
+            final IndexMetadata indexMetadata = state.getMetadata().index(index);
+
+            if (indexMetadata == null) {
+                // Index not found
+                listener.onFailure(new IndexNotFoundException(request.index));
+                return;
+            }
+
+            final int numberOfShards = indexMetadata.getNumberOfShards();
```

Comment:
It does work currently with wait for advance = true and multi shard? I understand that is not a use case, but it feels weird to make it throw an exception when it works fine now and we eventually want to support multi shard.

Reply:
I think it will result in two problems:

1. Not responding in a timely manner if one shard is advanced but the others are not.
2. Throw a timeout exception once the timeout period has passed, if only one shard advanced.

Maybe I am wrong about that?

Reply:
I guess up to you. I understand that the multi-shard functionality is less valuable than the single shared behavior currently. The areas you identify are outlined in the meta issue to be fixed in the next follow-up PRs. It just feels weird to pull out validation and tests when they are just going to come back in the next PRs.